### ISSUE TITLE
[DISCUSSION] Đánh giá hyperparameters hiện tại, gợi ý điều chỉnh, và đề xuất refactor cấu trúc code cho dễ debug

### ISSUE THREAD

**Issue by 23520761** [23520761/23520761.github.io#7]
title: "[DISCUSSION] Đánh giá hyperparameters hiện tại, gợi ý điều chỉnh, và đề xuất refactor cấu trúc code cho dễ debug"
repository: "23520761/23520761.github.io"
assignees:
  - "23520761"
labels:
  - "enhancement"
  - "refactor"
  - "discussion"
type: "task"
tag: "hyperparam-refactor-structure"
description: |
  ## 1. Đánh giá các hyperparameters hiện tại

  **Các giá trị hiện tại trong utils.py:**
  - BUFFER_SIZE = 10000
  - BATCH_SIZE = 64
  - NUM_IL_TRAJECTORIES = 100
  - NUM_IL_EPOCHS = 100
  - REPLAY_BUFFER_CAPACITY = 10000
  - TARGET_UPDATE_FREQ = 100
  - ALPHA = 0.001 (learning rate)
  - GAMMA = 0.99 (discount)
  - THRESHOLD = 0.6 (IoU threshold)
  - MAX_STEPS = 200
  - TRIGGER_STEPS = 10
  - NUMBER_ACTIONS = 6
  - ACTION_HISTORY_SIZE = 7
  - N_CLASSES = 20
  - TARGET_SIZE = (448, 448)
  - FEATURE_DIM = 512
  - EPOCHS = 100

  **Nhận xét & gợi ý điều chỉnh:**
  - BUFFER_SIZE, REPLAY_BUFFER_CAPACITY: 10k là hợp lý cho dataset vừa, có thể tăng lên nếu RAM đủ, giảm nếu máy yếu hoặc muốn debug nhanh vòng lặp.
  - BATCH_SIZE: 64 phù hợp, nếu GPU yếu có thể giảm còn 32, còn nếu GPU mạnh có thể tăng lên 128 để train nhanh hơn.
  - NUM_IL_TRAJECTORIES: 100 là khá ít, khuyến nghị tăng lên 500-1000 hoặc dùng hết batch train để IL pretrain ổn định hơn, đặc biệt khi số lượng class lớn.
  - NUM_IL_EPOCHS, EPOCHS: 100 hợp lý, tuy nhiên nên log loss để dễ kiểm soát convergence, nếu thấy loss dừng giảm thì dùng early stopping.
  - TARGET_UPDATE_FREQ: 100 chuẩn cho DQN, nếu thấy RL không ổn định, có thể giảm xuống 50 để target mạng cập nhật nhanh hơn.
  - ALPHA: 0.001 thích hợp với Adam, nếu model sâu hơn hoặc train lâu có thể giảm còn 0.0005.
  - THRESHOLD: 0.6 phù hợp chuẩn object detection, nếu muốn agent dễ hơn có thể hạ xuống 0.5.
  - MAX_STEPS: 200 hợp lý, nếu muốn episode ngắn hơn để debug có thể giảm 100-150.
  - Các thông số khác (actions, history, n_classes, feature_dim, target_size) đồng bộ với đề xuất và mô hình hiện tại.
  - **Khuyến nghị:** Nên log các thông số này khi chạy và tune theo thực nghiệm thực tế, nhất là NUM_IL_TRAJECTORIES và THRESHOLD.

  **Tóm lại:** Thông số hiện tại khá ổn cho VOC/COCO, chú ý tăng số trajectory IL, log kỹ các chỉ số convergence/training để dễ kiểm soát chất lượng.

  ---

  ## 2. Gợi ý refactor code & cấu trúc thư mục

  **Vấn đề:** File env.py lên đến 600 dòng, imitation_learning.py, agents.py... cũng dài, gây khó maintain/debug, khó mở rộng.

  **Đề xuất cấu trúc lại thư mục dự án:**

  ```
  project_root/
  ├── main.py
  ├── config.py                  # Hyperparameters, config chung (hoặc config.yaml)
  ├── utils/
  │   ├── __init__.py
  │   ├── replay_buffer.py       # ReplayBuffer và các tiện ích buffer
  │   ├── metrics.py             # mAP, IoU, recall, ...
  │   ├── transforms.py          # transform_input, augmentation, preprocessing
  │   └── visualization.py       # Các hàm plot, hiển thị ảnh, vẽ bbox
  ├── env/
  │   ├── __init__.py
  │   ├── base_env.py            # Lớp base Env chung, tiền xử lý
  │   ├── detection_env.py       # DetectionEnv (center/size logic)
  │   ├── reward.py              # Các hàm reward function, reward shaping
  │   ├── render.py              # Các hàm render/display, display debug
  │   └── utils.py               # Hàm phụ trợ liên quan env
  ├── il/
  │   ├── __init__.py
  │   ├── trajectory.py          # Sinh trajectory IL, sinh expert data
  │   ├── train_il.py            # Train model IL (center & size agent)
  │   ├── policy.py              # Chính sách expert, IL policy
  │   └── utils.py               # Loss function, logging IL
  ├── dqn/
  │   ├── __init__.py
  │   ├── agent.py               # Center/size agent RL
  │   ├── train_dqn.py           # RL pipeline training
  │   ├── models.py              # Q-network, head, backbone
  │   └── utils.py               # Loss, epsilon scheduling, logging RL
  ├── dataset/
  │   ├── __init__.py
  │   ├── voc.py                 # Pascal VOC loader
  │   └── coco.py                # COCO loader nếu mở rộng
  ├── test.py
  ├── requirements.txt
  └── README.md
  ```

  **Cách các file/module gọi nhau:**
  - main.py: orchestrate pipeline, gọi env, agent, train/test, load config
  - config.py: lưu hyperparams, import dùng chung các module
  - env/detection_env.py: chỉ giữ logic đặc thù, không chứa quá nhiều hàm phụ/mã render
  - utils/: các hàm phụ trợ, metric, transform, replay buffer, visualization
  - il/: quản lý toàn bộ imitation learning (trajectory, train, policy, loss)
  - dqn/: quản lý RL agent, Q-network, train RL, epsilon schedule
  - dataset/: loader từng loại dataset, split, augmentation, chuẩn hóa

  **Lợi ích:**
  - Dễ debug, dễ mở rộng, dễ test đơn vị từng module riêng
  - Có thể import lại từng module nhỏ mà không lo vòng lặp import hoặc xung đột namespace
  - Tăng maintainability, chia code rõ theo chức năng, thuận tiện cho teamwork, dễ refactor về sau

  **Gợi ý thêm:**
  - Dùng logging thay vì print, log ra file riêng (log/debug.log)
  - Toàn bộ hyperparameter nên gom vào config.py hoặc config.yaml, truyền vào main.py
  - Viết test đơn vị cho các hàm core (reward, iou, split, ...), có thể đặt trong tests/
  - Đặt tên biến và tên hàm rõ ý nghĩa, hạn chế dùng biến tắt, comment đầu hàm/class
  - Chia nhỏ hàm dài thành nhiều hàm nhỏ, tách logic phức tạp sang utils.py/module phụ
  - Giữ các file < 300 dòng; env.py, agents.py,... nếu dài hơn thì chia module như cấu trúc trên

  ---

  **Đề nghị:**
  - Xem xét tăng trajectory IL, tune lại threshold nếu cần cho phù hợp dataset.
  - Tiến hành tách nhỏ các module lớn, refactor code thành các package/module như gợi ý để tiện maintain, debug và phát triển lâu dài.
